api/request_processor.go

```diff
@@ -0,0 +1,350 @@
+		ReadBufferSize:  2048,
+		WriteBufferSize: 2048,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+)
+
+type RequestProcessor struct {
+	sessionManager mc.SessionManager
+	gameManager    mb.GameManager
+	dbManager      sqlc.DbManager
+	ipnet          net.IPNet
+}
+
+func NewRequestProcessor(
+	sessionManager mc.SessionManager,
+	gameManager mb.GameManager,
+	dbManager sqlc.DbManager,
+) RequestProcessor {
+	rp := RequestProcessor{
+		sessionManager: sessionManager,
+		gameManager:    gameManager,
+		dbManager:      dbManager,
+	}
+
+	rp = rp.mustGetServerIpNet()
+	return rp
+}
+
+func (rp RequestProcessor) mustGetServerIpNet() RequestProcessor {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		panic(err)
+	}
+
+	for _, iface := range ifaces {
+		// If the flag is down
+		if iface.Flags&net.FlagUp == 0 {
+			continue
+		}
+
+		if iface.Flags&net.FlagLoopback != 0 {
+			continue
+		}
+
+		addrs, err := iface.Addrs()
+		if err != nil {
+			panic(err)
+		}
+
+		for _, addr := range addrs {
+			var ipnet *net.IPNet
+			var ip net.IP
+
+			switch v := addr.(type) {
+			case *net.IPNet:
+				ipnet = v
+				ip = v.IP
+
+			case *net.IPAddr:
+				ip = v.IP
+			}
+
+			if ip != nil && ip.To4() != nil && !ip.IsLoopback() {
+				rp.ipnet = *ipnet
+				return rp
+			}
+		}
+	}
+
+	panic("ipnet could not be found!")
+}
+
+// Expose this method to use it in testing
+func (rp RequestProcessor) GetIpNet() net.IPNet {
+	return rp.ipnet
+}
+
+
+func (rp RequestProcessor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// use Upgrade method to make a websocket connection
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "could not open websocket connection", http.StatusBadRequest)
+		return
+	}
+
+	sessionIdQuery := r.URL.Query().Get(URLQuerySessionIDKeyword)
+	switch sessionIdQuery {
+	case "":
+		log.Println("a new connection established\tRemote Addr: ", conn.RemoteAddr().String())
+		rp.processSessionRequests(rp.sessionManager.GenerateNewSession(conn))
+
+	default:
+		session, err := rp.sessionManager.FindSession(sessionIdQuery)
+		if err != nil {
+			// This either means an expired session or invalid session ID
+			conn.WriteJSON(mc.NewMessage[mc.NoPayload](mc.CodeReceivedInvalidSessionID))
+			conn.Close()
+			return
+		}
+
+		rp.sessionManager.ReconnectSession(session, conn)
+		/*
+			we discussed that if app total closure or crash happens
+			it is not the server's fault. Hence, the server doesn not need
+			to provide the session information upon reconnection
+			Send the session data to update client information
+		*/
+	}
+}
+
+func (sp *RequestProcessor) processSessionRequests(session *mc.Session) {
+	defer sp.sessionManager.TerminateSession(session)
+
+	var (
+		otherSessionPlayer mb.Player
+		sessionPlayer      mb.Player
+		sessionGame        *mb.Game
+
+		receiverSessionId string
+		sessionId         = sp.sessionManager.GetSessionId(session)
```

Comment:
This could be written like `session.id` instead of `sp.sessionManager.GetSessionId(session)` ?